### PR TITLE
bench: add serializtion bench case

### PR DESCRIPTION
**Pull request by lishihai9017**
@BusyJay @zhangjinpeng1987 @siddontang PTAL

### CODE REVIEW

File: benches/serialization/bench_serialization.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    let value = Vec::from(String::from("0650bfba52").as_bytes());
+    r.mut_put().set_value(value);
+    reqs.push(r);
+
+    let mut r = Request::new();
+    r.set_cmd_type(CmdType::Put);
+    r.mut_put().set_cf(String::from("tidb"));
+    let key = Vec::from(String::from("6ba952020fbc91bad64be1ea0650bfba52e6aab4").as_bytes());
+    r.mut_put().set_key(key);
+    let value = Vec::from(String::from("6ba952020fbc91bad64be1ea0650bfba52e6aab4049b9e4e8067b998e4581d026b0bc6d1113ab9f507aaca3a0724000e735a558d4c23b600512346d9024aa9a345e92aa1926517c4d9b16bd83e74c10d6ba952020fbc91bad64be1ea0650bfba52e6aab4024aa9a345e92aa1926517c4d9b16bd83e74c10d7fd7b5b7fc628b3c").as_bytes());
```
Comment (zhangjinpeng1987) [inline benches/serialization/bench_serialization.rs]: Please use a const variable.

File: benches/serialization/bench_serialization.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    print!("{}", s);
+    print!(" cost {}ms\t...\t", cost);
+}
+
+#[bench]
+fn bench_encode_one_cmd(b: &mut Bencher) {
+    let count = 1000000;
+    let now = Instant::now();
+    for _ in 0..count {
+        encode_one_cmd();
```
Comment (zhangjinpeng1987) [inline benches/serialization/bench_serialization.rs]: Please extract `encode_one_cmd` and `encode_two_cmd` to a common function.

File: benches/serialization/bench_serialization.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    e.set_data(cmd_msg);
+    protobuf::Message::write_to_bytes(&e).unwrap()
+}
+
+fn decode(data: &Vec<u8>) {
+    let entry = protobuf::parse_from_bytes::<Entry>(data).unwrap();
+    let cmd = protobuf::parse_from_bytes::<RaftCmdRequest>(entry.get_data());
+}
+
+fn print_cost(s: &str, d: Duration, count: u32) {
```
Comment (siddontang) [inline benches/serialization/bench_serialization.rs]: no need to use this function, bench will output the statistics later.

File: benches/serialization/bench_serialization.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    let cost = (d.as_secs() * 1000) as f64 + (d.subsec_nanos() / 1000_000) as f64;
+    let cost = cost / count as f64;
+    print!("{}", s);
+    print!(" cost {}ms\t...\t", cost);
+}
+
+#[bench]
+fn bench_encode_one_cmd(b: &mut Bencher) {
+    let count = 1000000;
```
Comment (siddontang) [inline benches/serialization/bench_serialization.rs]: use bench iterator, don't use count

File: benches/serialization/bench_serialization.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+#[inline]
+fn encode_two_cmd() -> Vec<u8> {
+    let mut e = Entry::new();
+    let mut cmd = RaftCmdRequest::new();
+    let mut reqs = vec![];
+
+    let mut r = Request::new();
+    r.set_cmd_type(CmdType::Put);
+    r.mut_put().set_cf(String::from("tidb"));
+    let key = Vec::from(String::from("7fd7b5b7fc628b3cd19c56daf84dbe").as_bytes());
```
Comment (siddontang) [inline benches/serialization/bench_serialization.rs]:
```
let key = b"7fd7b5b7fc628b3cd19c56daf84dbe"
```

File: benches/serialization/mod.rs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (BusyJay) [inline benches/serialization/mod.rs]: s/2016/2017

File: benches/benches.rs [hunk 7]
```diff
@@ -26,8 +26,13 @@ extern crate mio;
 extern crate rand;
 extern crate tikv;
 extern crate time;
+extern crate tempdir;
```
Comment (BusyJay) [inline benches/benches.rs]: Why link `tempdir`?

File: benches/benches.rs [hunk 8]
```diff
@@ -26,8 +26,13 @@ extern crate mio;
...
+extern crate protobuf;
+extern crate kvproto;
+extern crate rocksdb;
```
Comment (BusyJay) [inline benches/benches.rs]: Why link rocksdb?

File: benches/serialization/bench_serialization.rs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate protobuf;
+
+use test::Bencher;
+use kvproto::eraftpb::Entry;
+use kvproto::raft_cmdpb::{RaftCmdRequest, Request, CmdType};
+use std::string::String;
+use std::collections::HashMap;
```
Comment (andelf) [inline benches/serialization/bench_serialization.rs]: reorder imports

File: benches/serialization/bench_serialization.rs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    let mut cmd = RaftCmdRequest::new();
+    let cloned_reqs = reqs.clone();
+    cmd.set_requests(protobuf::RepeatedField::from_vec(cloned_reqs));
+    let cmd_msg = protobuf::Message::write_to_bytes(&cmd).unwrap();
+    e.set_data(cmd_msg);
+    protobuf::Message::write_to_bytes(&e).unwrap()
+}
+
+fn decode(data: &Vec<u8>) {
+    let entry = protobuf::parse_from_bytes::<Entry>(data).unwrap();
```
Comment (andelf) [inline benches/serialization/bench_serialization.rs]: &[u8]

File: benches/serialization/bench_serialization.rs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        r.set_cmd_type(CmdType::Put);
+        r.mut_put().set_cf(String::from("tikv"));
+        r.mut_put().set_key(key.to_vec());
+        r.mut_put().set_value(value.to_vec());
+        reqs.push(r);
+    }
+    reqs
+}
+
+fn encode(reqs: &Vec<Request>) -> Vec<u8> {
```
Comment (andelf) [inline benches/serialization/bench_serialization.rs]: &[Request]

File: benches/serialization/bench_serialization.rs [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+fn decode(data: &[u8]) {
+    let entry = protobuf::parse_from_bytes::<Entry>(data).unwrap();
+    protobuf::parse_from_bytes::<RaftCmdRequest>(entry.get_data()).unwrap();
+}
+
+#[bench]
+fn bench_encode_one(b: &mut Bencher) {
+    let mut map: HashMap<&[u8], &[u8]> = HashMap::new();
+    map.insert(b"6ba952020fbc91bad64be1ea0650bf", b"6ba952020fbc91bad64be1ea0650bfba52e6aab4049b9e4e8067b998e4581d026b0bc6d1113ab9f507aaca3a0724000e735a558d4c23b600512346d9024aa9a345e92aa1926517c4d9b16bd83e74c10d6ba952020fbc91bad64be1ea0650bfba52e6aab4024aa9a345e92aa1926517c4d9b16bd83e74c10d7fd7b5b7fc628b3c");
```
Comment (siddontang) [inline benches/serialization/bench_serialization.rs]: The value is too long, you can use `vec![0:n]` directly.

**Comment by siddontang** [comment]
Please show us the benchmark output here.

**Comment by siddontang** [comment]
Rest LGTM

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

File: benches/serialization/bench_serialization.rs [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+fn bench_decode_two(b: &mut Bencher) {
+    let mut map: HashMap<&[u8], &[u8]> = HashMap::new();
+    map.insert(b"7fd7b5b7fc628b3cd19c56daf84dbe", b"6ba952020fbc91bad64be1ea0650bfba52e6aab4049b9e4e8067b998e4581d026b0bc6d1113ab9f507aaca3a0724000e735a558d4c23b600512346d9024aa9a345e92aa1926517c4d9b16bd83e74c10d6ba952020fbc91bad64be1ea0650bfba52e6aab4024aa9a345e92aa1926517c4d9b16bd83e74c10d7fd7b5b7fc628b3c");
+    map.insert(b"6ba952020fbc91bad64be1ea0650bf", b"6ba952020fbc91bad64be1ea0650bfba52e6aab4049b9e4e8067b998e4581d026b0bc6d1113ab9f507aaca3a0724000e735a558d4c23b600512346d9024aa9a345e92aa1926517c4d9b16bd83e74c10d6ba952020fbc91bad64be1ea0650bfba52e6aab4024aa9a345e92aa1926517c4d9b16bd83e74c10d7fd7b5b7fc628b3c");
+    let reqs = generate_requests(&map);
+    let data = encode(&reqs);
+    b.iter(|| {
+        decode(data.as_slice());
+    });
+}
```
Comment (andelf) [inline benches/serialization/bench_serialization.rs]: EOF

File: benches/benches.rs [hunk 14]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016 PingCAP, Inc.
+// Copyright 2017 PingCAP, Inc.
```
Comment (siddontang) [inline benches/benches.rs]: 2016

**Comment by lishihai9017** [comment]
## Serialization Result
```
test bench_decode_one ... bench:         686 ns/iter (+/- 95)
test bench_decode_two ... bench:       1,275 ns/iter (+/- 360)
test bench_encode_one ... bench:         486 ns/iter (+/- 93)
test bench_encode_two ... bench:         787 ns/iter (+/- 196)
```

**Comment by siddontang** [comment]
seems that we should try to reduce the key count.

File: benches/serialization/bench_serialization.rs [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+extern crate protobuf;
+
+use std::string::String;
+use std::collections::HashMap;
+use test::Bencher;
+use rand;
+use kvproto::eraftpb::Entry;
+use kvproto::raft_cmdpb::{RaftCmdRequest, Request, CmdType};
+
+fn gen_rand_str(len: usize) -> Vec<u8> {
```
Comment (BusyJay) [inline benches/serialization/bench_serialization.rs]: Use `Rng::fill_bytes` or `rand::Generator` instead.

File: benches/serialization/bench_serialization.rs [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    }
+    bstr
+}
+
+fn generate_requests(map: &HashMap<&[u8], &[u8]>) -> Vec<Request> {
+    let mut reqs = vec![];
+    for (key, value) in map {
+        let mut r = Request::new();
+        r.set_cmd_type(CmdType::Put);
+        r.mut_put().set_cf(String::from("tikv"));
```
Comment (BusyJay) [inline benches/serialization/bench_serialization.rs]: Use `"tikv".to_owned()` instead.

File: benches/serialization/bench_serialization.rs [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+        reqs.push(r);
+    }
+    reqs
+}
+
+fn encode(reqs: &[Request]) -> Vec<u8> {
+    let mut e = Entry::new();
+    let mut cmd = RaftCmdRequest::new();
+    cmd.set_requests(protobuf::RepeatedField::from_vec(reqs.to_vec()));
+    let cmd_msg = protobuf::Message::write_to_bytes(&cmd).unwrap();
```
Comment (BusyJay) [inline benches/serialization/bench_serialization.rs]: Use `cmd.write_to_bytes()` instead.

File: benches/serialization/bench_serialization.rs [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+    e.set_data(cmd_msg);
+    protobuf::Message::write_to_bytes(&e).unwrap()
+}
+
+fn decode(data: &[u8]) {
+    let entry = protobuf::parse_from_bytes::<Entry>(data).unwrap();
```
Comment (BusyJay) [inline benches/serialization/bench_serialization.rs]: Use `merge_from` instead.

File: benches/serialization/bench_serialization.rs [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+
+#[bench]
+fn bench_encode_one(b: &mut Bencher) {
+    let key = gen_rand_str(30);
+    let value = gen_rand_str(256);
+    let mut map: HashMap<&[u8], &[u8]> = HashMap::new();
+    map.insert(key.as_slice(), value.as_slice());
+    let reqs = generate_requests(&map);
+    b.iter(|| {
+        encode(reqs.as_slice());
```
Comment (BusyJay) [inline benches/serialization/bench_serialization.rs]: s/reqs.as_slice/&reqs/

File: benches/serialization/bench_serialization.rs [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+        r.mut_put().set_value(value.to_vec());
+        reqs.push(r);
+    }
+    reqs
+}
+
+fn encode(reqs: &[Request]) -> Vec<u8> {
+    let mut e = Entry::new();
+    let mut cmd = RaftCmdRequest::new();
+    cmd.set_requests(protobuf::RepeatedField::from_vec(reqs.to_vec()));
```
Comment (BusyJay) [inline benches/serialization/bench_serialization.rs]: Does the copy effect the bench result?

**Comment by siddontang** [comment]
any update? @lishihai9017

File: benches/serialization/bench_serialization.rs (hunk 18 above)
Comment (lishihai9017) [inline benches/serialization/bench_serialization.rs]: no such method found

File: benches/serialization/bench_serialization.rs (hunk 17 above)
Comment (lishihai9017) [inline benches/serialization/bench_serialization.rs]: no such method found

Comment (siddontang) [inline benches/serialization/bench_serialization.rs]:
```
use protobuf::Message
```

File: benches/serialization/bench_serialization.rs (hunk 20 above)
Comment (lishihai9017) [inline benches/serialization/bench_serialization.rs]: yes, it's cost about 150ns.

**Comment by BusyJay** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**